cloudinit/config/cc_set_hostname.py

```diff
@@ -21,9 +21,13 @@
 the ``fqdn`` config key. If both ``fqdn`` and ``hostname`` are set, ``fqdn``
 will be used.
 
+In order to correctly set the hostname, cloud-init will always run this module
```

Comment:
I'm not sure this is exactly true. This is the place that it's called during `main` execution:

https://github.com/canonical/cloud-init/blob/7765aac637a757c736674b0eb2e2993c01442dbc/cloudinit/cmd/main.py#L705-L709

so we can see that there is some conditonal logic based on `hostname`. `util.get_hostname_fqdn` is what sets `hostname`, and its docstring says https://github.com/canonical/cloud-init/blob/7765aac637a757c736674b0eb2e2993c01442dbc/cloudinit/util.py#L1105-L1106

We _do_ pass `metadata_only=True`, so there could be cases when we _don't_ run `cc_set_hostname` module during init.

All that said, I don't think I understand well enough when we wouldn't run it to suggest different wording. Perhaps something like:

```suggestion
In order to correctly set the hostname, cloud-init will always run this module if there is metadata specifying a host name
```

but I'm not sure that really clarifies matters much. :-/

Reply:
What do you think about this block (also needs wordsmithing):


In order to correctly set the hostname, cloud-init will always run this module
int the init-local stage before networking is configured if metadata or
userdata from a local source (e.g. nocloud, ovf) specifies a hostname.
Some platforms base dynamic DNS on intial hostnames requested via DHCP
requests. This behavior ensures that any desired hostname set from metadata or
userdata is applied before any DHCP requests are performed on the machine.

Setting a hostname in init-local stage only happens if the following conditions are met:
 * local metadata provides a `local-hostname` value **OR** user-data or
   system config in /etc/cloud/cloud.cfg.d/*.cfg provides a `hostname` value
    **AND**
 * the current hostname of the VM does not match the metadata or userdata
   hostname found above